nucypher/blockchain/eth/token.py

```diff
@@ -275,14 +282,134 @@ def validate_duration(self, raise_on_fail=True) -> Union[bool, Tuple[Tuple[bool,
+    #
+
+    @classmethod
+    def __deposit(cls, miner, amount: int, lock_periods: int) -> Tuple[str, str]:
+        """Public facing method for token locking."""
+
+        approve_txhash = miner.token_agent.approve_transfer(amount=amount,
+                                                            target_address=miner.miner_agent.contract_address,
+                                                            sender_address=miner.checksum_public_address)
+
+        deposit_txhash = miner.miner_agent.deposit_tokens(amount=amount,
+                                                          lock_periods=lock_periods,
+                                                          sender_address=miner.checksum_public_address)
+
+        return approve_txhash, deposit_txhash
+
+    def divide(self, target_value: NU, additional_periods: int = None) -> tuple:
+        """
+        Modifies the unlocking schedule and value of already locked tokens.
+
+        This actor requires that is_me is True, and that the expiration datetime is after the existing
+        locking schedule of this miner, or an exception will be raised.
+       """
+
+        # Re-read stakes from blockchain and select stake to divide
+        current_stake = self
+
+        # Ensure selected stake is active
+        if current_stake.is_expired:
+            raise self.StakingError(f'Cannot divide an expired stake')
+
+        if target_value >= current_stake.value:
+            raise self.StakingError(f"Cannot divide stake; Target value ({target_value}) must be less "
+                                    f"than the existing stake value {current_stake.value}.")
+
+        #
+        # Generate SubStakes
+        #
+
+        # Modified Original Stake
+        modified_stake = Stake(miner=self.miner,
+                               index=self.index,
+                               start_period=current_stake.start_period,
+                               end_period=current_stake.end_period,
+                               value=target_value)
+
+        # New Derived Stake
+        end_period = current_stake.end_period + additional_periods
+        new_stake = Stake(miner=self.miner,
+                          start_period=current_stake.start_period,
+                          end_period=end_period,
+                          value=current_stake.value - target_value,
+                          index=NEW_STAKE)
+
+        #
+        # Validate
+        #
+
+        # Ensure both halves are for valid amounts
+        modified_stake.validate_value()
+        new_stake.validate_value()
+
+        # Detect recycled or inconsistent slot then check start period or fail
+        stake_info = self.miner_agent.get_stake_info(miner_address=self.owner_address, stake_index=self.index)
+        first_period, last_period, locked_value = stake_info
+        if not modified_stake.start_period == first_period:
+            raise self.StakingError("Inconsistent staking cache, aborting stake division. ")
```

Comment:
Should place this block in the beginning of the method (after `current_stake == self`). And after this check `current_stake` = `stake_info`. Because if stake was slashed then `current_stake.value` and `locked_value` will be different (and even `end_period`/`last_period` can be different)

Reply:
:heavy_check_mark: